opentracing-spring-messaging/src/main/java/io/opentracing/contrib/spring/integration/messaging/OpenTracingChannelInterceptor.java

```diff
@@ -0,0 +1,168 @@
+ * the License.
+ */
+
+package io.opentracing.contrib.spring.integration.messaging;
+
+import io.opentracing.ActiveSpan;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.propagation.Format;
+import io.opentracing.tag.Tags;
+import java.util.Collections;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.springframework.integration.channel.AbstractMessageChannel;
+import org.springframework.integration.context.IntegrationObjectSupport;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.MessageHandler;
+import org.springframework.messaging.support.ChannelInterceptorAdapter;
+import org.springframework.messaging.support.ExecutorChannelInterceptor;
+import org.springframework.util.ClassUtils;
+
+/**
+ * @author <a href="mailto:gytis@redhat.com">Gytis Trikleris</a>
+ */
+public class OpenTracingChannelInterceptor extends ChannelInterceptorAdapter implements ExecutorChannelInterceptor {
+
+  private static final Log log = LogFactory.getLog(OpenTracingChannelInterceptor.class);
+
+  private static final String COMPONENT_TAG = "spring-messaging";
+
+  private static final String MESSAGE_COMPONENT = "message";
+
+  private final Tracer tracer;
+
+  public OpenTracingChannelInterceptor(Tracer tracer) {
+    this.tracer = tracer;
+  }
+
+  @Override
+  public Message<?> preSend(Message<?> message, MessageChannel channel) {
+    log.trace("Processing message before sending it to the channel");
+
+    MessageTextMap<?> carrier = new MessageTextMap<>(message);
+    SpanContext parentSpan = tracer.extract(Format.Builtin.TEXT_MAP, carrier);
+    String operationName = getOperationName(channel);
+    ActiveSpan span = tracer.buildSpan(operationName)
+        .asChildOf(parentSpan)
+        .startActive();
+
+    Tags.COMPONENT.set(span, COMPONENT_TAG);
+    Tags.MESSAGE_BUS_DESTINATION.set(span, getChannelName(channel));
+
+    if (message.getHeaders()
+        .containsKey(Headers.MESSAGE_SENT_FROM_CLIENT)) {
+      log.trace("Marking span with server received");
+      span.log(Events.SERVER_RECEIVE);
+      span.setBaggageItem(Events.SERVER_RECEIVE, Events.SERVER_RECEIVE);
+      Tags.SPAN_KIND.set(span, Tags.SPAN_KIND_CONSUMER);
+      // TODO maybe we should remove Headers.MESSAGE_SENT_FROM_CLIENT header here?
+    } else {
+      log.trace("Marking span with client send");
+      span.log(Events.CLIENT_SEND);
+      Tags.SPAN_KIND.set(span, Tags.SPAN_KIND_PRODUCER);
+      carrier.put(Headers.MESSAGE_SENT_FROM_CLIENT, "true");
+    }
+
+    tracer.inject(span.context(), Format.Builtin.TEXT_MAP, carrier);
+
+    return carrier.getMessage();
+  }
+
+  @Override
+  public void afterSendCompletion(Message<?> message, MessageChannel channel, boolean sent, Exception ex) {
+    ActiveSpan activeSpan = tracer.activeSpan();
```

Comment:
@pavolloffay what's the usual way to get a span here? Should I try to extract it from the tracer first and then use activeSpan()?